File: Comportamentales Java/Interpreter.java
// Interpreter Pattern
interface AbstractExpression {
    boolean interpret(String context);
}

class TerminalExpression implements AbstractExpression {
    private String data;
    
    public TerminalExpression(String data) {
        this.data = data;
    }
    
    @Override
    public boolean interpret(String context) {
        return context.contains(data);
    }
}

class OrExpression implements AbstractExpression {
    private AbstractExpression expr1;
    private AbstractExpression expr2;
    
    public OrExpression(AbstractExpression expr1, AbstractExpression expr2) {
        this.expr1 = expr1;
        this.expr2 = expr2;
    }
    
    @Override
    public boolean interpret(String context) {
        return expr1.interpret(context) || expr2.interpret(context);
    }
}

class AndExpression implements AbstractExpression {
    private AbstractExpression expr1;
    private AbstractExpression expr2;
    
    public AndExpression(AbstractExpression expr1, AbstractExpression expr2) {
        this.expr1 = expr1;
        this.expr2 = expr2;
    }
    
    @Override
    public boolean interpret(String context) {
        return expr1.interpret(context) && expr2.interpret(context);
    }
}

// Usage
class InterpreterDemo {
    public static void main(String[] args) {
        AbstractExpression robert = new TerminalExpression("Robert");
        AbstractExpression john = new TerminalExpression("John");
        AbstractExpression isMale = new OrExpression(robert, john);
        
        AbstractExpression julie = new TerminalExpression("Julie");
        AbstractExpression married = new TerminalExpression("Married");
        AbstractExpression isMarriedWoman = new AndExpression(julie, married);
        
        String context1 = "John";
        String context2 = "Married Julie";
        
        System.out.println("John is male? " + isMale.interpret(context1));
        System.out.println("Julie is a married woman? " + isMarriedWoman.interpret(context2));
    }
}

File: Comportamentales Python/interpreter.py
# Interpreter Pattern
from abc import ABC, abstractmethod

class AbstractExpression(ABC):
    @abstractmethod
    def interpret(self, context):
        pass

class TerminalExpression(AbstractExpression):
    def __init__(self, data):
        self.data = data
    
    def interpret(self, context):
        return self.data in context

class OrExpression(AbstractExpression):
    def __init__(self, expr1: AbstractExpression, expr2: AbstractExpression):
        self.expr1 = expr1
        self.expr2 = expr2
    
    def interpret(self, context):
        return self.expr1.interpret(context) or self.expr2.interpret(context)

class AndExpression(AbstractExpression):
    def __init__(self, expr1: AbstractExpression, expr2: AbstractExpression):
        self.expr1 = expr1
        self.expr2 = expr2
    
    def interpret(self, context):
        return self.expr1.interpret(context) and self.expr2.interpret(context)

# Usage
if __name__ == "__main__":
    robert = TerminalExpression("Robert")
    john = TerminalExpression("John")
    is_male = OrExpression(robert, john)
    
    julie = TerminalExpression("Julie")
    married = TerminalExpression("Married")
    is_married_woman = AndExpression(julie, married)
    
    context1 = "John"
    context2 = "Married Julie"
    
    print(f"John is male? {is_male.interpret(context1)}")
    print(f"Julie is a married woman? {is_married_woman.interpret(context2)}")
